File: 카카오프렌즈컬러링북/jane.java
import java.util.*;

class Solution {

    private static boolean[][] visited;
    private static final int[] dx = {-1, 0, 1, 0};
    private static final int[] dy = {0, 1, 0, -1};

    static class Pair {
        int x;
        int y;

        Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public int bfs(int x, int y, int[][] picture, int m, int n) {
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(x, y));
        visited[x][y] = true;
        int sum = 1;

        while (!queue.isEmpty()) {
            Pair pair = queue.remove();

            int nextX, nextY;
            for (int i = 0; i < 4; i++) {
                nextX = pair.x + dx[i];
                nextY = pair.y + dy[i];

                if (nextX >= 0 && nextY >= 0 && nextX < m && nextY < n && !visited[nextX][nextY]
                        && picture[pair.x][pair.y] == picture[nextX][nextY]) {
                    visited[nextX][nextY] = true;
                    queue.add(new Pair(nextX, nextY));
                    sum++;
                }
            }
        }
        return sum;
    }

    public int[] solution(int m, int n, int[][] picture) {
        int numberOfArea = 0;
        int maxSizeOfOneArea = 0;
        visited = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && picture[i][j] != 0) {
                    maxSizeOfOneArea = Math.max(maxSizeOfOneArea, bfs(i, j, picture, m, n));
                    numberOfArea++;
                }
            }
        }

        int[] answer = new int[2];
        answer[0] = numberOfArea;
        answer[1] = maxSizeOfOneArea;
        return answer;
    }
}

File: 카카오프렌즈컬러링북/jane.py
from collections import deque

dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]


def solution(m, n, picture):
    number_of_area = 0
    max_size_of_one_area = 0
    global visited

    visited = [[False] * n for _ in range(m)]

    for i in range(m):
        for j in range(n):
            # 방문 표시도 안 되었고 그림에 색도 칠해져 있을 때
            if (not visited[i][j]) and picture[i][j] != 0:
                max_size_of_one_area = max(max_size_of_one_area, bfs(i, j, picture, m, n))
                number_of_area += 1
    return number_of_area, max_size_of_one_area


def bfs(i, j, picture, m, n):
    count = 1
    pixel = picture[i][j]
    queue = deque()
    queue.append((i, j))

    visited[i][j] = True
    while queue:
        y, x = queue.popleft()
        for idx in range(4):
            nextY, nextX = y + dy[idx], x + dx[idx]
            if 0 <= nextY < m and 0 <= nextX < n and not visited[nextY][nextX] and pixel == picture[nextY][nextX]:
                queue.append((nextY, nextX))
                visited[nextY][nextX] = True
                count += 1
    return count


print(solution(6, 4, [[1, 1, 1, 0], [1, 2, 2, 0], [1, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 3], [0, 0, 0, 3]]))
print(solution(6, 4, [[1, 1, 1, 0], [1, 1, 1, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1]]))
